### ISSUE TITLE
[java] UnusedPrivateMethod false positive

### ISSUE THREAD

**Issue by anicoara** [pmd/pmd#1187]
**Affects PMD Version:** Maven PMD 3.10

**Rule:** UnusedPrivateMethod 

**Description:** When running Maven PMD plugin the UnusedPrivateMethod rule is hit if a derived object is passed as an argument to a private method taking a base type argument. I can't reproduce this with inner classes, but it works with classes that are in a different package than the one that is using them.

**Code Sample demonstrating the issue:**

Project structure:

```
├── pmd-ruleset.xml
├── pom.xml
└── src
    └── main
        └── java
            └── test
                ├── App.java
                └── foo
                    ├── A.java
                    └── B.java
```

Contents of `A.java`:

```java
package test.foo;

public class A
{
}
```

Contents of `B.java`:

```java
package test.foo;

public final class B extends A
{
}
```

Contents of `App.java`:

```java
package test;

import test.foo.A;
import test.foo.B;

public class App
{
    private void foo(A a)
    {
    }

    public void bar()
    {
        final B b = new B();
        //foo((A)b); // passes
        foo(b);
    }
}
```

Error:

```
[INFO] PMD Failure: test.App:8 Rule:UnusedPrivateMethod Priority:3 Avoid unused private methods such as 'foo(A)'..
```

**Running PMD through:** *[Maven]*

Note, running PMD 6.4.0 from the CLI passes:

```
pmd -d . -R pmd-ruleset.xml 
```

**Comment by adangel** [comment]
@anicoara Thanks for the report.
How do you execute the maven-pmd-plugin exactly? The problem stems from the difference, when typeresolution (aka auxclasspath) is in use or not. The maven pmd plugin by default enables typeresolution (configuration option [typeresolution](http://maven.apache.org/plugins/maven-pmd-plugin/pmd-mojo.html#typeResolution)), whereas on the command line, you would need to specify [auxclasspath](https://pmd.github.io/pmd-6.4.0/pmd_userdocs_getting_started.html) explicitly.
And now, it depends, whether the project is compiled or not - the referenced classes are in the same project and PMD can only know about it after compilation.

In short: If you just execute `mvn pmd:check`, this may yield false-positives. It should work, if you execute `mvn compile pmd:check` or you just execute `mvn verify`.

Could you please try this and leave a comment, whether this solves the problem?

**Comment by anicoara** [comment]
I see - if I disable `typeResolution` it indeed works. The issue is that I am running pmd during the `validate` phase, so I can't compile if I can't pass that phase.

I read [this similar bug](https://sourceforge.net/p/pmd/bugs/1468/) and I think I am starting to grasp the need for `typeResolution` to be turned on, but then I am not sure how to do the checks before compiling.

**Comment by anicoara** [comment]
Maybe moving the check to the `test` phase is more appropriate. I will close the issue, as it seems it was my misuse.

Thank you for the help.

**Comment by adangel** [comment]
Yes, it is usually necessary to compile the project first to get the best results with PMD.
The [check goal](http://maven.apache.org/plugins/maven-pmd-plugin/check-mojo.html) of the maven-pmd-plugin automatically binds to the verify phase, which is after the compilation is done for the project.
